File: 0015-3sum/solution.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {

        Arrays.sort(nums);

        List<List<Integer>> res = new ArrayList<>();
        int n = nums.length;

        for (int i = 0; i < n; i++) {
            if (i > 0 && nums[i - 1] == nums[i])
                continue;

            int first = nums[i];

            int p = i + 1;
            int q = nums.length - 1;

            while (p < q) {
                if (nums[p] + nums[q] + first == 0) {
                    res.add(List.of(nums[p], nums[q], first));
                    p++;
                    q--;

                    while (p < q && nums[p - 1] == nums[p])
                        p++;
                    while (p < q && nums[q + 1] == nums[q])
                        q--;
                } else if (nums[p] + nums[q] + first < 0) {
                    p++;
                } else {
                    q--;
                }
            }
        }

        return res;
    }
}

File: 0015-3sum/solution.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        # deduplicate, return num
        nums.sort()
        res = []

        for i in range(len(nums)):
            if nums[i] > 0:
                break

            if i > 0 and nums[i] == nums[i - 1]:
                continue

            p, q = i + 1, len(nums) - 1

            while p < q:
                if nums[i] + nums[p] + nums[q] < 0:
                    p += 1
                elif nums[i] + nums[p] + nums[q] > 0:
                    q -= 1
                else:
                    res.append([nums[i], nums[p], nums[q]])
                    p += 1
                    q -= 1

                    while p < q and nums[p] == nums[p - 1]:
                        p += 1

                    while p < q and nums[q + 1] == nums[q]:
                        q -= 1

        return res

